src/main/java/uk/nhs/tis/sync/job/reval/RevalCurrentPmSyncJob.java

```diff
@@ -0,0 +1,105 @@
+
+/**
+ * Get personIds whose current programmeMembership changes nightly.
+ * And sends messages to rabbitMq for tcs to fetch
+ */
+@Component
+@ManagedResource(objectName = "sync.mbean:name=RevalCurrentPmSyncJob",
+    description = "Job message personIds if their programme membership(s) started/ended")
+public class RevalCurrentPmSyncJob extends PersonCurrentPmSyncJobTemplate {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RevalCurrentPmSyncJob.class);
+
+  private final EntityManagerFactory entityManagerFactory;
+  private final RabbitMqTcsPmUpdatePublisher rabbitMqPublisher;
+
+  public RevalCurrentPmSyncJob(EntityManagerFactory entityManagerFactory,
+      RabbitMqTcsPmUpdatePublisher rabbitMqPublisher) {
+    this.entityManagerFactory = entityManagerFactory;
+    this.rabbitMqPublisher = rabbitMqPublisher;
+  }
+
+  @Override
+  public void run(String params) {
+    revalCurrentPmSyncJob();
+  }
+
+  @Scheduled(cron = "${application.cron.revalCurrentPmJob}")
+  @SchedulerLock(name = "revalCurrentPmScheduledTask", lockAtLeastFor = FIFTEEN_MIN,
+      lockAtMostFor = FIFTEEN_MIN)
+  @ManagedOperation(
+      description = "send personIds to tcs for reval current programmeMembership sync")
+  public void revalCurrentPmSyncJob() {
+    runSyncJob(null);
+  }
+
+  protected EntityManagerFactory getEntityManagerFactory() {
+    return entityManagerFactory;
+  }
+
+  protected String getDateOfChangeOverride() {
+    return null;
+  }
+
+  protected void doDataSync(String queryString) {
+    int totalRecords = 0;
+    long lastEntityId = 0;
+    boolean hasMoreResults = true;
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+
+    EntityManager entityManager = null;
+    try {
+      while (hasMoreResults) {
+        entityManager = getEntityManagerFactory().createEntityManager();
+
+        List<Long> collectedData =
+            collectData(lastEntityId, queryString, entityManager);
+        hasMoreResults = !collectedData.isEmpty();
+        LOG.info("Time taken to read chunk : [{}]", stopwatch);
+
+        if (CollectionUtils.isNotEmpty(collectedData)) {
+          rabbitMqPublisher.publishToBroker(
+              collectedData.stream().map(String::valueOf).collect(
+                  Collectors.toList()));
+          lastEntityId = collectedData.get(collectedData.size() - 1);
+          totalRecords += collectedData.size();
+        }
+        stopwatch.reset().start();
+
+        entityManager.close();
+        stopwatch.reset().start();
+      }
+      LOG.info("Sync job [{}] finished. Total time taken {} for processing [{}] records",
+          getJobName(), mainStopWatch.stop(), totalRecords);
+      mainStopWatch = null;
+      publishJobexecutionEvent(
+          new JobExecutionEvent(this, getSuccessMessage(Optional.ofNullable(getJobName()))));
+    } finally {
+      mainStopWatch = null;
+
+      if (entityManager != null && entityManager.isOpen()) {
+        entityManager.close();
+      }
+    }
+  }
```

Comment:
Alot of this looks like what is in the `PersonRecordStatusJob.doDataSync`.  It looks like you've identified the `convertData()` step is different and I think there is a "persist" step that could be put in as an delegated method too.

Reply:
Thanks Pepe. 
In the new commits, I refactored this bit. I've moved this whole process into the superclass, and implemented abstract methods `convertData` & `handleData` in child classes. For the new job, it doesn't really any data conversion, but should match this patter of process. I hope this refactoring looks better and makes more sense.